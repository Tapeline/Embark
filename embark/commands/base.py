from abc import ABC, abstractmethod
from argparse import Namespace
from typing import Any, final

from attrs import frozen

from embark.domain.interfacing.installs_provider import InstallationsInterface
from embark.domain.interfacing.os_provider import OSInterface


@final
@frozen
class NoArgs:
    """Blank args class."""


class AbstractCommand[_ArgsT](ABC):
    """ABC for all commands."""

    _args_type: type[_ArgsT]

    def __init__(
            self,
            os_interface: OSInterface[InstallationsInterface[Any]]
    ) -> None:
        """Create command and link os interface."""
        self.os_interface = os_interface

    def __call__(self, args: Namespace) -> int:
        """Parse arguments into the generic namespace."""
        args_dict = vars(args)  # noqa: WPS421
        if "func" in args_dict:
            args_dict.pop("func")
        cmd_args = self._args_type(**args_dict)
        return self._run(cmd_args)

    @abstractmethod
    def _run(self, args: _ArgsT) -> int:
        """Run the command."""
        raise NotImplementedError
