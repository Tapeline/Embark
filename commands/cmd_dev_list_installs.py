"""
dev_list_installs subcommand
"""


# pylint: disable=unused-argument
def command(args):
    """Subcommand impl"""
    # pylint: disable=import-outside-toplevel
    from std.target.install.installs_repo import WindowsInstallsRepository
    repo = WindowsInstallsRepository()
    for i, install in enumerate(repo.get_all_installs()):
        print(f"--- {i + 1} ---")
        print(f"Name:      {install.name}")
        print(f"Version:   {install.version}")
        print(f"Publisher: {install.publisher}")
