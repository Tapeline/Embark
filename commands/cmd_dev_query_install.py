"""
dev_query_install subcommand
"""
import re


def command(args):
    """Subcommand impl"""
    # pylint: disable=import-outside-toplevel
    from std.target.install.installs_repo import WindowsInstallsRepository
    repo = WindowsInstallsRepository()
    for install in repo.get_all_installs():
        if (re.fullmatch(args.name, install.name) is not None
                and (install.publisher == args.publisher or install.publisher is None) and
                (install.version == args.version or args.ignore_version)):
            print(f"Name:      {install.name}")
            print(f"Version:   {install.version}")
            print(f"Publisher: {install.publisher}")
            return
