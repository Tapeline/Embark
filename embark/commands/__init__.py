"""This module contains all Embark CLI commands."""
