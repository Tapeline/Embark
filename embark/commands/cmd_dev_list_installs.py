"""dev_list_installs subcommand."""

from embark.std.target.install.installs_repo import WindowsInstallsRepository


def command(args):
    """Subcommand impl."""
    repo = WindowsInstallsRepository()
    for i, install in enumerate(repo.get_all_installs()):
        print(f"--- {i + 1} ---")
        print(f"Name:      {install.name}")
        print(f"Version:   {install.version}")
        print(f"Publisher: {install.publisher}")
