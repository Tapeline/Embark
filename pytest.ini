[pytest]
addopts =
    --cov-report=term-missing:skip-covered
    --cov-report=html:htmlcov
    --cov-report=xml
    --cov=embark
    --cov=tests
