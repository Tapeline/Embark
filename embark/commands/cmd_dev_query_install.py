"""dev_query_install subcommand."""

from typing import final

from attrs import frozen

from embark.commands.base import AbstractCommand
from embark.output import write_out


@final
@frozen
class DevQueryInstallCommandArgs:
    """Arguments for dev_query_install command."""

    name: str
    publisher: str
    version: str
    ignore_version: bool = False


class DevQueryInstallCommand(AbstractCommand[DevQueryInstallCommandArgs]):
    """dev_query_install command."""

    _args_type = DevQueryInstallCommandArgs

    def _run(self, args: DevQueryInstallCommandArgs) -> int:
        """Subcommand impl."""
        repo = self.os_interface.get_install_interface()
        got_any = False
        for install in repo.get_all_installs():
            if install.matches(
                args.name,
                args.publisher,
                args.version,
                ignore_version=args.ignore_version
            ):
                got_any = True
                write_out(f"Name:      {install.name}")
                write_out(f"Version:   {install.version}")
                write_out(f"Publisher: {install.publisher}")
        return 0 if got_any else 1
