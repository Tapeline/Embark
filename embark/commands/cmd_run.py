"""run subcommand."""

import time

from embark.application import Application


def command(args):
    """Subcommand impl"""
    app = Application(args.filename, file_encoding=args.encoding)
    app.run()
    time.sleep(0.5)
    input("Press [ENTER] to close")  # noqa: WPS421
