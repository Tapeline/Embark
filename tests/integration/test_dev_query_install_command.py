from argparse import Namespace

import pytest

from embark.commands.cmd_dev_query_install import DevQueryInstallCommand
from tests.integration.mock import MockOSInterface, mock_streams


@pytest.mark.parametrize(
    ("name", "publisher", "version"),
    [
        ("NameA", "PubA", "1.0"),
        ("NameB", "PubA", ...),
        (".*", "PubB", "1.2"),
        (".*", ".*", ...)
    ]
)
def test_dqi_command(name, publisher, version, snapshot):
    """Test that embark dqi works."""
    with mock_streams() as streams:
        command = DevQueryInstallCommand(MockOSInterface())
        ret_code = command(Namespace(
            name=name,
            publisher=publisher,
            version=None if version is ... else version,
            ignore_version=version is ...
        ))
    assert ret_code == 0
    assert streams.stdout.value == snapshot
    assert not streams.stderr.value
