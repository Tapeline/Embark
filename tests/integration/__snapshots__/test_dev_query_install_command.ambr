# serializer version: 1
# name: test_dqi_command[.*-.*-version3]
  '''
  Name:      NameA
  Version:   1.0
  Publisher: PubA
  Name:      NameB
  Version:   1.1
  Publisher: PubA
  Name:      NameC
  Version:   1.2
  Publisher: PubB
  Name:      NameD
  Version:   1.3
  Publisher: PubB
  
  '''
# ---
# name: test_dqi_command[.*-PubB-1.2]
  '''
  Name:      NameC
  Version:   1.2
  Publisher: PubB
  
  '''
# ---
# name: test_dqi_command[NameA-PubA-1.0]
  '''
  Name:      NameA
  Version:   1.0
  Publisher: PubA
  
  '''
# ---
# name: test_dqi_command[NameB-PubA-version1]
  '''
  Name:      NameB
  Version:   1.1
  Publisher: PubA
  
  '''
# ---
